### PR TITLE
support for multiple files in ModelCheckpoint

### PR DESCRIPTION
**Pull request by amitbeka**
enable string formatted filenames (e.g. weights.{epoch:02d}.hdf5), so
every epoch will be saved to a different file without overwriting.

Signed-off-by: Amit Beka <amit.beka@gmail.com>

### CODE REVIEW

**Comment by wuaalb** [comment]
I think this is a good idea.

It might be also nice to add an option to ModelCheckpoint to allow saving only every n-th epoch, instead of every epoch (to avoid using up too much disk space when each epoch has a different filename).

**Comment by kendler** [comment]
Hi,

maybe that helps: I had the same problem and simply made my own callback which allows saving every "EpochDistance". I added the epoch and the loss in the name of the file. So there is no overwriting.
```
import keras
class MyModelCheckpoint(keras.callbacks.Callback):
    def __init__(self, filepath, EpochDistance=100):
        super(keras.callbacks.Callback, self).__init__()
        self.filepath = filepath
        self.EpochDistance= EpochDistance
    def on_train_begin(self, logs={}):
        self.history = {}
    def on_epoch_begin(self, epoch, logs={}):
        self.seen = 0
        self.totals = {}
    def on_batch_end(self, batch, logs={}):
        batch_size = logs.get('size', 0)
        self.seen += batch_size
        for k, v in logs.items():
            if k in self.totals:
                self.totals[k] += v * batch_size
            else:
                self.totals[k] = v * batch_size
    def on_epoch_end(self, epoch, logs={}):
        for k, v in self.totals.items():
            if k not in self.history:
                self.history[k] = []
            self.history[k] = (v / self.seen)
        for k, v in logs.items():
            if k not in self.history:
                self.history[k] = []
            self.history[k] = v
        if np.mod(epoch, self.EpochDistance) == 0:
            print("Epoch %05d: saving model to %s" % (epoch, self.filepath+'_Epoch_'+str(epoch)+'_loss_'+str(round(self.history.values()[1],5))+'.h5py'))
            self.model.save_weights(self.filepath+'_Epoch_'+str(epoch)+'_loss_'+str(round(self.history.values()[1],5))+'.h5py', overwrite=True)
```

**Comment by amitbeka** [comment]
You can really create custom callback with more sophsiticated logic like you did. I thought it may be nice not to override-by-design all checkpoints, but allow multiple ones, with minimal code addition (it's basically a one-line change, calling `format(*args, **kwargs)` on the filename).

@fchollet -- what's your opinion on this patch?

**Comment by fchollet** [comment]
> what's your opinion on this patch?

Looks good to me, I think this is a useful feature to have. Merged. Sorry for the delay in reviewing this.
